internal/xds/resolver/xds_http_filters_test.go

```diff
@@ -0,0 +1,678 @@
+	"google.golang.org/grpc/internal/testutils/xds/e2e"
+	"google.golang.org/grpc/internal/xds/httpfilter"
+	rinternal "google.golang.org/grpc/internal/xds/resolver/internal"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/structpb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	v3xdsxdstypepb "github.com/cncf/xds/go/xds/type/v3"
+	v3clusterpb "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	v3endpointpb "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	v3listenerpb "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	v3routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	v3httppb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	testgrpc "google.golang.org/grpc/interop/grpc_testing"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
+
+	_ "google.golang.org/grpc/xds" // Register all required xDS components
+)
+
+const (
+	filterCfgPathFieldName  = "path"
+	filterCfgErrorFieldName = "new_stream_error"
+	filterCfgMetadataKey    = "test-filter-config"
+)
+
+// testFilterCfg is the internal representation of the filter config proto. It
+// is returned by filter's config parsing methods.
+type testFilterCfg struct {
+	httpfilter.FilterConfig
+	path         string
+	newStreamErr error
+}
+
+// filterConfigFromProto parses filter config specified as a v3.TypedStruct into
+// a testFilterCfg.
+func filterConfigFromProto(cfg proto.Message) (httpfilter.FilterConfig, error) {
+	ts, ok := cfg.(*v3xdsxdstypepb.TypedStruct)
+	if !ok {
+		return nil, fmt.Errorf("unsupported filter config type: %T, want %T", cfg, &v3xdsxdstypepb.TypedStruct{})
+	}
+
+	if ts.GetValue() == nil {
+		return testFilterCfg{}, nil
+	}
+	ret := testFilterCfg{}
+	if v := ts.GetValue().GetFields()[filterCfgPathFieldName]; v != nil {
+		ret.path = v.GetStringValue()
+	}
+	if v := ts.GetValue().GetFields()[filterCfgErrorFieldName]; v != nil {
+		if v.GetStringValue() == "" {
+			ret.newStreamErr = nil
+		} else {
+			ret.newStreamErr = fmt.Errorf("%s", v.GetStringValue())
+		}
+	}
+	return ret, nil
+}
+
+type logger interface {
+	Logf(format string, args ...any)
+}
+
+// testHTTPFilterWithRPCMetadata is a HTTP filter used for testing purposes.
+//
+// This filter is used to verify that the xDS resolver and filter stack
+// correctly propagate filter configuration (both base and override) to RPCs. It
+// does this by injecting the config paths from its base and override configs as
+// JSON-encoded metadata into outgoing RPCs.  The metadata can then be observed
+// by the backend, allowing tests to assert that the correct filter
+// configuration was applied for each RPC.
+type testHTTPFilterWithRPCMetadata struct {
+	logger        logger
```

Comment:
nit: We can directory store a `testing.T` here instead of abstracting it using an interface.

Reply:
It's probably an overkill, but I feel that it's better to do this instead of accepting a `testing.T`, since we can very easily start calling other methods on it like `Error` or `Fatal` if we have access to it.